### ISSUE TITLE
[BUG] crc start fails at INFO Waiting for user's pull secret part of instance disk...

### ISSUE THREAD

**Issue by adithyagk** [code-ready/crc#2872]
### General information

  * OS: Linux 
  * Hypervisor: KVM
  * Did you run `crc setup` before starting it (Yes/No) Yes
  * Running CRC on: Laptop 

## CRC version
```bash
# Put `crc version` output here

`CodeReady Containers version: 1.36.0+c0f4e0d3
OpenShift version: 4.9.8 (embedded in executable)`
```
  
## CRC status
```bash
# Put `crc status --log-level debug` output here

`[adithya@localhost ~]$ crc status --log-level debug
DEBU CodeReady Containers version: 1.36.0+c0f4e0d3 
DEBU OpenShift version: 4.9.8 (embedded in executable) 
DEBU Running 'crc status'                         
DEBU Checking file: /home/adithya/.crc/machines/crc/.crc-exist 
DEBU Checking file: /home/adithya/.crc/machines/crc/.crc-exist 
DEBU Found binary path at /home/adithya/.crc/bin/crc-driver-libvirt 
DEBU Launching plugin server for driver libvirt   
DEBU Plugin server listening at address 127.0.0.1:36868 
DEBU () Calling .GetVersion                       
DEBU Using API Version 1                          
DEBU () Calling .SetConfigRaw                     
DEBU () Calling .GetMachineName                   
DEBU (crc) Calling .GetState                      
DEBU (crc) DBG | time="2021-12-03T12:05:58+05:30" level=debug msg="Getting current state..." 
DEBU (crc) DBG | time="2021-12-03T12:05:58+05:30" level=debug msg="Fetching VM..." 
DEBU (crc) Calling .GetBundleName                 
DEBU (crc) Calling .GetIP                         
DEBU (crc) DBG | time="2021-12-03T12:06:00+05:30" level=debug msg="GetIP called for crc" 
DEBU (crc) DBG | time="2021-12-03T12:06:00+05:30" level=debug msg="Getting current state..." 
DEBU (crc) DBG | time="2021-12-03T12:06:00+05:30" level=debug msg="IP address: 192.168.130.11" 
DEBU Running SSH command: df -B1 --output=size,used,target /sysroot | tail -1 
DEBU Using ssh private keys: [/home/adithya/.crc/machines/crc/id_ecdsa /home/adithya/.crc/cache/crc_libvirt_4.9.8/id_ecdsa_crc] 
DEBU SSH command results: err: <nil>, output: 32737570816 13687865344 /sysroot 
DEBU cannot get OpenShift status: stat /home/adithya/.crc/machines/crc/kubeconfig: no such file or directory 
DEBU Making call to close driver server           
DEBU (crc) Calling .Close                         
DEBU Successfully made call to close driver server 
DEBU Making call to close connection to plugin binary 
CRC VM:          Running
OpenShift:       Unreachable (v4.9.8)
Disk Usage:      13.69GB of 32.74GB (Inside the CRC VM)
Cache Usage:     15.6GB
Cache Directory: /home/adithya/.crc/cache
`
```

## CRC config
```bash
# Put `crc config view` output here

`[adithya@localhost ~]$ crc config view
- consent-telemetry                     : yes
- pull-secret-file                      : /home/adithya/Downloads/pull-secret.txt
`
```

## Host Operating System
```bash
# Put the output of `cat /etc/os-release` in case of Linux
`[adithya@localhost ~]$ cat /etc/os-release
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
`
```

### Steps to reproduce

  1. Just running crc start

### Expected

openshift cluster running

### Actual
failing at
`INFO Waiting for kube-apiserver availability... [takes around 2min] 
INFO Adding user's pull secret to the cluster...  
INFO Updating SSH key to machine config resource... 
INFO Waiting for user's pull secret part of instance disk... 
Failed to update kubeadmin user password: Temporary error: ssh command error:
command : timeout 5s oc get secret --context admin --cluster crc --kubeconfig /opt/kubeconfig
err     : Process exited with status 124
 (x12)`

### Logs

Before gather the logs try following if that fix your issue
```bash
$ crc delete -f
$ crc cleanup
$ crc setup
$ crc start --log-level debug
```

Please consider posting the output of `crc start --log-level debug`  on http://gist.github.com/ and post the link in the issue.
https://gist.github.com/adithyagk/4f7319d49cac498e119c125f166e9911

**Comment by praveenkumar** [comment]
@adithyagk Are you able to reproduce it everytime after `crc delete` => `crc start` ?  As part of our code base same pull secret should be part of VM disk image, can you try to login the VM https://github.com/code-ready/crc/wiki/Debugging-guide and try to check the content of `/var/lib/kubelet/config.json` file?

**Comment by adithyagk** [comment]
hi @praveenkumar , This was happening since morning, and yes its reproduced everytime i did crc delete => start.
I will post the update after i login to the VM. Iam doing the same process all over so it will take 10 mins.

**Comment by adithyagk** [comment]
hi @praveenkumar , it looks like the token similar to the one i provided for pull secret

**Comment by cfergeau** [comment]
Is this running on a physical host? Or are you running crc in CentOS7, which itself is running in a VM? (vmware, ..?)

**Comment by adithyagk** [comment]
I have a laptop which is running CentOS7, its in a dual boot fashion with windows 10

**Comment by adithyagk** [comment]
Hi @praveenkumar @cfergeau  is there any configuration that i should do from my end to resolve this?

**Comment by praveenkumar** [comment]
@adithyagk Can you try `crc stop` and then `crc start` again, if it taking bit more time also try to increase the resources like `memory` to the instance. How much ram does your host have?

**Comment by adithyagk** [comment]
i have just tried crc stop => crc delete => crc start
I have 16gigs of Ram

**Comment by adithyagk** [comment]
Hi Guys. i was able to resolve this, not sure if its a temporary fix.
I installed another release of openshift https://developers.redhat.com/content-gateway/rest/mirror/pub/openshift-v4/clients/crc/1.33.1.
Did the same process again, set 14000 memory.
Still failed at the exact same mark. Proceeded to login to the VM. Checked the /var/lib/kubelet/config.json file, this time it was empty (the error was still the same), edited it with the token i received, gave crc start again.
Able to login to console now, However iam able to login as developer from UI and not with admin credentials, i dont know if the problem is associated with it

**Comment by adithyagk** [comment]
Hi, i just verified, iam unable to login with admin credentials from CLI as well with the admin credentials provided.

crc status still shows openshift as unreachable.

`[adithya@localhost ~]$ crc status
CRC VM:          Running
OpenShift:       Unreachable (v4.8.12)
Disk Usage:      14.71GB of 32.74GB (Inside the CRC VM)
Cache Usage:     16.06GB
Cache Directory: /home/adithya/.crc/cache`

**Comment by adithyagk** [comment]
This is however not working with the crc containers for 4.9version, just checked it. Kindly do let me know if there is any proper solution for this

**Comment by adithyagk** [comment]
Hi, 
Just another observation, i removed CentOs and installed Fedora35,
Same error happened, 
For the first time on Crc start, i can see that /var/lib/kubelet/config.json is empty.
Even after that i run crc start again and iam prompted to
```bash
[adithya@fedora ~]$ crc start
INFO A CodeReady Containers VM for OpenShift 4.9.8 is already running 
Started the OpenShift cluster.

The server is accessible via web console at:
  https://console-openshift-console.apps-crc.testing

Log in as administrator:
  Username: kubeadmin
  Password: <somepassword>

Log in as user:
  Username: developer
  Password: developer

Use the 'oc' command line interface:
  $ eval $(crc oc-env)
  $ oc login -u developer https://api.crc.testing:6443
```

The link however doesnt load .

I logged into VM and updated /var/lib/kubelet/config.json manually with my token.

Ran Crc start again .
Got the same output as quoted before, this time iam able reach the web console.

The only catch is that i can only access as a developer and not kubeadmin hence i cannot do anything admin related.

I have shared all the findings from my side, kindly guide me on what to do, if any alterations in the config can solve this as a workaround similar to what i have done to update the token which resulted in the web console to work.

I want to login as kubeadmin and it doesnt seem to be possible with the current approach as i get 401 unauthorized

**Comment by adithyagk** [comment]
Another observation (Hopefully its the issue)

I think the time-out happens a bit too early.
Observation:-

crs start resulted in the following error:
```bash
INFO Waiting for kube-apiserver availability... [takes around 2min] 
Error waiting for apiserver: Temporary error: ssh command error:
command : timeout 5s oc get nodes --context admin --cluster crc --kubeconfig /opt/kubeconfig
err     : Process exited with status 1
 (x27)
Temporary error: ssh command error:
command : timeout 5s oc get nodes --context admin --cluster crc --kubeconfig /opt/kubeconfig
err     : Process exited with status 124
 (x3)
Temporary error: ssh command error:
command : timeout 5s oc get nodes --context admin --cluster crc --kubeconfig /opt/kubeconfig
err     : Process exited with status 1
 (x7)
Temporary error: ssh command error:
command : timeout 5s oc get nodes --context admin --cluster crc --kubeconfig /opt/kubeconfig
err     : Process exited with status 124
 (x3)
Temporary error: ssh command error:
command : timeout 5s oc get nodes --context admin --cluster crc --kubeconfig /opt/kubeconfig
err     : Process exited with status 1
 (x21)
Temporary error: ssh command error:
command : timeout 5s oc get nodes --context admin --cluster crc --kubeconfig /opt/kubeconfig
err     : Process exited with status 124
 (x21)
```
In a separate terminal i ssh into the VM and ran the same command, i didnt get any output for like 15-20 seconds, after which i got the following:
```bash
[core@crc-x4qnm-master-0 ~]$ oc get nodes --context admin --cluster crc --kubeconfig /opt/kubeconfig
NAME                 STATUS   ROLES           AGE   VERSION
crc-x4qnm-master-0   Ready    master,worker   9d    v1.22.2+5e38c72
```
I would like to believe that crc start times out a bit before the actual output of 
`oc get nodes --context admin --cluster crc --kubeconfig /opt/kubeconfig`
gets into affect.

Also kubeadmin isnt added as well which i believe is because crc start ends prematurely .
```bash
[core@crc-x4qnm-master-0 ~]$ KUBECONFIG=/opt/kubeconfig kubectl get users
NAME        UID                                    FULL NAME   IDENTITIES
developer   <someid>               developer:developer
```

**Comment by praveenkumar** [comment]
@adithyagk If it is timeout issue then I would suggest to try following. I am still not sure why it is time out, also try to close all other process when running the crc if it is caused by some IO issue.

```
$ crc delete
$ crc start => let it fails prematurely
$ crc stop => Don't delete again just stop it.
$ crc start => Hopefully this will provide extra time for secret and other content
```

**Comment by adithyagk** [comment]
Hi @praveenkumar  I tried the above approach before and the console view pops up, but again the kubeadmin credential doesnt get generated and i wwont be able to login as kubeadmin.

As for current, i tried it with centos, RHEL 7.9 and fedora 35, all report the same issue( to rule out the possibility of OS version causing the issue)

As a last resort ( to check if my comp is the culprit or not, whether its 4vcpu causing the issue) i logged into a VM in google cloud => created a vm with 8CPU cores and 30 GB memory nested virtualization capable centos 7 , then ran the setup, this time it ran successfully without no issues.
To get the console i had to install xrdp and RDP from my windows machine. But in the end everything works great!!

In short i believe the minimum requirement is an understatement, it would probably require like minimum 6c CPU to work( I may be wrong here)

**Comment by adithyagk** [comment]
Hi, iam closing this as i consider it an issue with the minimum system requirement.
Thanks for all the help!!

**Comment by m1820** [comment]
I had the same issue and it was very frustrating. I have the following configuration and this is how I was able to finally get crc installed:
- vCenter Server
- RHEL 8 with 8CPUs, 20GB RAM, 80GB drive

I did a few things and not 100% which one was exactly what fixed it but here are the 2 main thing I did a little different:

- Changed permissions on the host file `sudo chmod 666 /etc/hosts`
- I ran `crc start -c 6 -m 16000 -p pull-secret --log-level debug` 
The difference I noticed is that if I did not specify the CPU and RAM or if I used the exact amount the system had it will keep failing. The moment I specify 2 fewer CPUs and less RAM than what the system had, the installation was finally completed!!!
